src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java

```diff
@@ -195,49 +285,62 @@ public final boolean hasMore() throws NamingException {
      */
     @Override
     public final T next() throws NamingException {
-
-        if (!hasMoreCalled) {
-            hasMore();
+        try {
+            if (!hasMoreCalled) {
+                hasMore();
+            }
+            hasMoreCalled = false;
+            return nextImpl();
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
```

Comment:
The reachability fence should take 'this' as a parameter, not enumCtx. Since 'this' reachability is tracked by Cleaner and not 'enumCtx' reachability.